### ISSUE TITLE
SystemJS ♥️ unpkg

### ISSUE THREAD

**Issue by alexisvincent** [unpkg/express-unpkg#65]
I smell an opportunity here to provide a beautiful SystemJS loading experience with unpkg.

If you haven't already played with it, SystemJS is a client side module loader and can already provides a nice experience for module loading from cdns and unpkg (Angular does this in their official guides).

I think we could make the experience even nicer however by automatically resolving the configs for raw projects using [systemjs-config-builder](https://github.com/alexisvincent/systemjs-config-builder).

The experience on the browser side would look something like this:

`app.js`
```js
import { DOM } from 'react';
import { render } from 'react-dom';

render(
    DOM.div('Hello World'), 
    document.getElementById('app')
)
```

`index.html`
```html
<!DOCTYPE html>
<html>
<head>
  <title>SystemJS ♥️ unpkg</title>
  <script src="unpkg.com/systemjs@19.0.0/dist/system.js"></script>
  <script src="unpkg.com/systemjs@19.0.0/dist/unpkg-config.js"></script>
</head>
<body>
  <div id="app" />
  <script>
    System.import('./app.js')
  </script>
</body>
</html>
```

Which is obviously really attractive since the only dependency you need is SystemJS, which then automatically handles the rest. This makes barrier to entry for client side development really minimal. I could also write a hosted service which parsed client code and made loading modules in production efficient.

@mjackson Let me know if this is something you would be interested to support.

**Comment by alexisvincent** [comment]
Could even collapse
```html
  <script src="unpkg.com/systemjs@19.0.0/dist/system.js"></script>
  <script src="unpkg.com/systemjs@19.0.0/dist/unpkg-config.js"></script>
```
to
```html
  <script src="unpkg.com/systemjs-tools@1.3.8/unpkg.js"></script>
```

**Comment by mjackson** [comment]
Is the unpkg-config.js file something that we would generate on our server? In order to generate that file, would we need to do an npm install of the package?

**Comment by alexisvincent** [comment]
That config would be static. But we would need to generate other config files on the server after an npm install. But the config file could be cached just like the others

**Comment by mjackson** [comment]
What other config files would we need to generate on the server? Assume I know nothing about SystemJS. I'm having a hard time figuring out what would be required on our side.

**Comment by sebastien** [comment]
@alexisvincent I was wondering how you would manage the various pre-built JavaScript configurations we find on NPM projects. The main problem with NPM packages is that they're all implemented using a variety of precompilers (babel, traceur, tsc, etc) and when they ship with pre-built "true" JavaScript files (ie that would run as-is in the browser) these files can be placed and laid-out in many different ways.

For instance, https://unpkg.com/gl-matrix@2.3.2 is not usable as-is as the `require`s wont' resolve (`require("./gl-matrix/common.js") points to https://unpkg.com/gl-matrix@2.3.2/common.js. The other problem is that the JS files likely won't work as-is in the browser, and we should instead use the pre-built version in https://unpkg.com/gl-matrix@2.3.2/dist/gl-matrix.js. The https://unpkg.com/gl-matrix@2.3.2/package.json does not specify a `browser:` entry either, so how would you deal with this logic?

Having a per-package configuration would be unviable because of the quantity of NPM packages, so the only other way would be to implement a custom SystemJS resolver for unpkg that would get a directory listing as well as the `package.json` to identify where to pull the pre-built files from.

**Comment by alexisvincent** [comment]
Sorry for the delay in replying. I'm taking a few days off all work

**Comment by mjackson** [comment]
No worries, @alexisvincent :) We all need some time off

**Comment by alexisvincent** [comment]
Sorry that this has taken so long, I took a break and have been focusing on
other OSS work. I've given this some thought and have some new ideas.

I want to point out at the get go that there are 2 major areas that work in this direction would streamline. The first is user experience, the second is cache maximization. I will explore these two areas separately, but they work elegantly together.

Coupling a smart client side module loader (SystemJS) with an automatic dependency resolver (unpkg) could create a very compelling development AND production story.

### UX
The kind of workflow I'm envisioning is one where folks could seamlessly require packages and they would be automatically resolved from unpkg. A cli tool could also be provided to resolve the dependency tree upfront and provide a really solid production workflow.

### Cache
Apart from the user experience boon, we could actually gain a serious cache boost. Currently, the status quo is to download a bunch of dependencies with npm and then bundle them with your app into some big bundle. Then whenever we make changes to our app code, the whole bundle is invalidated. This story is made marginally better by progressively importing dependency trees as you need them, but still, there's often no barrier between app code (frequently changes) and dependencies (infrequently changes). Some people address this by bundling node_module dependencies separately to app code. While this is a step in the right direction, modern apps often have large transitive dependency trees and bundling dependencies as a whole means that whenever any one dep updates, the whole dep bundle is invalidated. Another problem is that creating your own bundles means that dep caches can't be shared across different applications.

The most cache efficient option is thus utilizing public, cdn cached builds. unpkg for example. The problem with that is that manual dependency management is a huge pain. It's the reason we invented package managers.

Let's assume for a moment that unpkg was able to provide browser friendly builds on a package level for all npm packages (explored further down). And instead of bundling dependencies together with your app you dynamically resolved them in the browser with a smart module loader. You get the best of both worlds, automatic dependency management as well as globally shared package level caching.

Enter SystemJS and unpkg.

Ideally, packages and our apps are actually resolved and cached on a file level, but until http2 and Service Workers are more widespread, and browsers improve parallel file parsing/loading performance, I think package level builds are a pretty good step forwards.

### Implementation
Ok, so there are a number of things that need to be built, and some more that would make the experience kickass.

#### Background
First, some background for folks unfamiliar with SystemJS and browser module loading in general. Assume no bundles or prebuilt files (they can be considered a convenience).

Loading in the browser is more nuanced then on the server, since we have to know up front exactly where files are, since the time to localise them experimentally is too high. Take for instance the case where a package author calls `require('./foo')`. What they implicitly mean is `./foo.js` and node correctly resolves this, since it checks if `./foo.js` exists. However, the author ***could*** have also meant `./foo.json`. The point is there's no way to resolve the filename upfront, and the performance cost of doing so in the browser is too high.
Another issue here, is that the node resolution logic is also indeterministic. An npm install could place a node package in any number of places. Yarn introduces deterministic install locations, but as an outsider, without access to the yarn.lock, I would have no way of knowing where to actually find the package.
Because of this indeterminism, to load packages through SystemJS, we need to statically analyse the package on the server and generate mappings and other hints, so that we make the job of loading the package deterministic.

#### SystemJS compatibility and unoptimized loader
Assuming we wanted to load files individually (not great performance, but still useful) through unpkg, we would need access to the statically generated config for the package. Assuming we have this, writing an unpkg SystemJS loader would be trivial. Usage in app code would then look something like this (loader's concern):
```js
// explicite version and loader
import react from 'react@1.2.3!unpkg'

// latest version and explicite loader
import react from 'react!unpkg'

// pre-generated config mapping (easy) or automatic resolution
import react from 'react'
```

This is quite clearly a UX boost for the developer (even if not particularly performant), since they don't have to think about bundling or local dependencies through npm.

All that needs to be done on the `unpkg` side is to perform the static analysis on the package when it first resolves it, and to the serve that config file. This is easy to do and I've already done most of the grunt work through my [`systemjs-config-builder`](https://github.com/alexisvincent/systemjs-config-builder) package, which would just need some simple modifications. I was lucky since I can piggy back off of the static analysis functions that JSPM uses internally.

#### Live module bundling and optimized loader
Now that we have a config file for the package, we could leverage [systemjs/builder](https://github.com/systemjs/builder)'s universal build capability to generate builds for the package in any format (`umd`, `amd`, `cjs`, `esm`, `global`...) we could conceivably generate builds 'on the fly' with something similar to @sebastien's `?as=umd` url extension. Now not only can unpkg serve individual package files, but it can also serve arbitrary builds of the package.

This also means we can efficiently load an app with package level caching, and the UX described in my previous point. If folks jumped on to this way of doing things (as they have unpkg itself), it could also significantly contribute to globally shared caches.

In the case that a package already provides its own builds, those can be used in place of building our own. This can be determined via thing like the `browser` key in the `package.json`.

This would provide an ideal development AND production workflow and I think as an offering could greatly ease on-boarding in the web. It would also be really usefull for example apps and usage examples, since people could experiment inline without setting up a large build chain.

#### unpackaged (package manager)
Once these basic primitives have been implemented, we could even provide a package manager (jokingly called unpackaged) which could simply do module resolution for you, so that compatible package versions could be resolved and the browser would also know which package builds to load upfront. So package version control for unpkg :)

#### Local Development Server
One of the annoying things about developing against a CDN is that you need to always be 'online'. If you want to work in a plane or somewhere without easy internet access, you're stuck unless you have a local copy.

Since unpkg is open source, folks could actually just run a local version and resolve packages off of it in development.

### Wrap up
Sorry for the long post but I thought it was worth putting all my thoughts on paper so that everyone is on the same page. 

Am really interested to hear your feedback.

**Comment by mjackson** [comment]
Thanks for all the details, @alexisvincent. This sounds very exciting! A few quick questions:

> All that needs to be done on the unpkg side is to perform the static analysis on the package when it first resolves it, and to the serve that config file.

That means we need to install all of that package's dependencies, right? Because `systemjs-config-builder` looks through the `node_modules` directory to build the config file?

We're currently just a simple file server, and we don't actually ever run `npm install` (or `yarn install`) on our servers, so this would be a fairly significant change for us. We'd need to make sure we can do this securely and carefully manage disk space.

> Now that we have a config file for the package, we could leverage systemjs/builder's universal build capability to generate builds

How does this affect package authors? Can they just use ES6 modules? Or do they have to use `SystemJS.import`, etc.?

**Comment by mjackson** [comment]
Tell me more about the package manager. Is that the piece that people use to generate the SystemJS config file they need in production?

Also, the development server. How would it get the package sources without a `node_modules` dir to pull from? Would we keep our own cache on disk?

**Comment by alexisvincent** [comment]
> That means we need to install all of that package's dependencies, right? Because systemjs-config-builder looks through the node_modules directory to build the config file?
>
> We're currently just a simple file server, and we don't actually ever run npm install (or yarn install) on our servers, so this would be a fairly significant change for us. We'd need to make sure we can do this securely and carefully manage disk space.

My bad, when I was scanning through the codebase I saw stuff like this
```js
      checkLocalCache(packageDir, (isCached) => {
        if (isCached)
          return next(packageDir) // Best case: we already have this package on disk.
```
and assumed you saved local copies of the package. I'm guessing you then just proxy through to the npm registry on a file level? How do you manage getting a hold of the files?

As it currently stands `systems-config-builder` would need a local copy of at least all the files in the package. It doesn't need the full dependency tree, but obviously would eventually need to parse all of them to be able to resolve all files.

> How does this affect package authors? Can they just use ES6 modules? Or do they have to use SystemJS.import, etc.?

Package authors wouldn't need to change anything with regards to how they normally publish. `SystemJS.import` is just for dynamic requires. Typically we only use `SystemJS.import` at the entry points to our application.

So this approach would work on all packages that `SystemJS` and `JSPM` currently support. There are certain packages that fail. For instance, some packages do the following:
```js
const namedModules = ['one', 'two', 'three'].map( num => require(`./some/${num}.js`) )
```
For obvious reasons this kind of dynamic require is really difficult to statically analyze (ESM's change this since `import` is static, so as a community we're moving away from this). In this case the `SystemJS`/`JSPM` community has generally submitted PR's to the author to turn this kind of thing into something easier to load. Luckily, the static config used to load the package can easily be extended with compatibility fixes for the package. For example

```js
SystemJS.config({
  packages: {
    'someProblematicPackage': {
        map: {
            './noExtention': './noExtention.js',
            './funckyRequireFile.js': '/myFunckyOverrideFile.js'
        }
    }
  }
})
```
 And while waiting for the author to update their package, the `SystemJS`/`JSPM` community normally publishes their fix to the [JSPM registry](https://github.com/jspm/registry).

Luckily these kinds of patches are relatively rare and we would be able to leverage that registry and the community effort that has gone into compatibility. To date, the registry has over 2265 commits, most of those are overrides for individual package versions that were problematic through the years.

> Tell me more about the package manager. Is that the piece that people use to generate the SystemJS config file they need in production?

Without the package manager tool thing, loading modules would work just fine, but `SystemJS` would import the packages on the fly, introducing latency into the process, this would be fine generally, but in development and production (for different reasons) you want your site to load as quickly as possible. To fix this we need to tell `SystemJS` (the browser) upfront all of the packages we need, then `SystemJS` could preemptively load them in parallel. People would also want to control package versions and have version clashes resolved by a tool that understands `semver`. This needn't be a `cli` tool, it could also be an online tool, but having it ready as a cli command would be useful. In either case, the `'package manager'` would just enhance the experience, it's not required.

> Also, the development server. How would it get the package sources without a node_modules dir to pull from? Would we keep our own cache on disk?

Again, my assumption was that `unpkg` had access to the local files. But regardless of how it worked, we could just run `npm-http-server` on our local machines with a local file cache and use whatever mechanism it uses to serve files. Folks could also, if they wanted to, run their own production server. Also this would defeat the purpose of the global cache, but they would get per package caching.

**Comment by mjackson** [comment]
> As it currently stands systems-config-builder would need a local copy of at least all the files in the package. It doesn't need the full dependency tree, but obviously would eventually need to parse all of them to be able to resolve all files.

Ok, that sounds doable. We currently do keep a local cache of the files in a package, but not its dependencies. We just grab the .tgz from npm, decompress it, and serve files out of that.

**Comment by alexisvincent** [comment]
Cool. How do you feel about the rest?

**Comment by mjackson** [comment]
I think it's a very compelling UX and I'd love to work with you on this. I've already claimed the [`unpkg` npm package](https://npmjs.org/package/unpkg) in anticipation of providing a CLI that could help people use unpkg, but hadn't fully thought through exactly what it might do until now.

My workflow is traditionally to just install all my dependencies locally and then use a bundler (webpack) to put everything together when I deploy, so I'll probably have some questions about how SystemJS does things as we go. But I think my "outsider" influence could also help us to make sure the experience is as seamless as possible for people who aren't familiar with SystemJS. We can build on SystemJS tools, but I think we can keep the new user experience really simple so that people who are just getting started on a new project won't have to know too much about the underlying infrastructure in order to use our tool.

Let's try to develop a roadmap. Where do you think we should start?

**Comment by sebastien** [comment]
I think @alexisvincent use case is very compelling, and alleviates the need for using NPM altogether, as long as the `import` resolver resolves to unpkg.

The main questions I would have are:

1) how do you generate the per-package configuration file for SystemJS?
2) how would they be retrieved and injected into SystemJS' configuration?
3) how do you build any given NPM package so that it becomes loadable by any modern browser?

I think there's a possibility to implement a resolver plug-in for SystemJS that would work in most cases (ie. where the NPM package includes build files) without having to do any significant change to unpkg, so I'm curious to see how you envisioned these 3 points.

**Comment by EricSimons** [comment]
@alexisvincent "To fix this we need to tell SystemJS (the browser) upfront all of the packages we need, then SystemJS could preemptively load them in parallel." <- since all of the packages req'd would be getting resolved from unpkg, wouldn't the max concurrent connections to same domain make the loading experience unreasonably slow for dev & especially prod? (Assuming no HTTP2)

**Comment by alexisvincent** [comment]
@EricSimons  :/ Damn, I hadn't thought about that... Good catch. 

As you mentioned, in `http2` this is a non issue. But you're right. In http 1.1 this would effectively introduce a latency in-between almost most requests... What is the max in most browsers? 8?

**Comment by alexisvincent** [comment]
@mjackson even with @EricSimons's revelation, I can still see tons of uses for this. Probably most notably the on demand bundling and on-boarding UX. It's really annoying though about the max concurrent connections. That effectively (unless I can think of a way to work around this somehow), means that, for production, this is a no go. I have a few ideas about how we can get around this, just need to flesh them out a bit and then I'll share. Out of interest, @mjackson what percentage of connections to `unpkg` use http2?

@sebastien To answer your questions
> how do you generate the per-package configuration file for SystemJS?

`systemjs-config-builder` piggybacks off of the install time analysis that `JSPM` performs. And effectively does this on a package level recursively through node_modules. I would need to do some trivial modifications to decouple the traversal logic from the generation step and then we could do this very simply.

> how would they be retrieved and injected into SystemJS' configuration?

via a custom `unpkg` loader.

> how do you build any given NPM package so that it becomes loadable by any modern browser?

we could leverage any of the existing builders, so effectively doing what people currently do manually, automatically. We definitely wouldn't be able to support all packages. For instance serverside packages would fail. But the ones that people currently use in the browser would be simple to package.

@mjackson the fact that you're unfamiliar with SystemJS is definitely helpful. 

With regards to the roadmap. 

The first thing to do would probably be to agree on an external facing API for loading the config file and bundles.

Next would be for me to build in support with `systemjs-config-builder` for per package analysis. The api would effectively be `generateConfig('path/to/package')`, so this would need to be easily integratable on the unpkg side. What would it take to implement this? We would also then need to do this on the fly based on a query param. Something like: `https://unpkg.com/react@2.2.2?file=systemjs-config.json`.

@mjackson I could today or tomorrow provide a dummy `getConfig` function to do the generation. Would you be able to sort out the `?file=systemjs-config.json` -> `getConfig` mapping?

As soon as this is done, I can write a simple loader to resolve packages (very slowly) from `unpkg`. But it could get the basic workflow down. The nice thing would be that almost everything from this point onwards would be changes to the implementation (read optimizations) and the external API would remain the same.

I think in parallel we could also look at way to make running `npm-http-server` as a dev server easy.

Then I think the next thing to do would be to implement the 'on the fly' module bundling. Something we could do to alleviate some of the max connections problem, is tie bundle granularity to the protocol. So for instance with `http2` we could serve the bundled package without deps. Then with http1, we could serve a package with its deps bundled in. Perhaps this isn't the best way to deal with it. But I'm confident we will find a viable solution. We would need to come up with a url extension that works and then tie that to the module bundling process.

This would allow me to expand the loader to load on a package level instead of a file level. I think aiming for this point is a good starting goal.

After that we would need the version resolution and upfront package loading and provide that probably via a `cli` and web tool.

@mjackson As a general roadmap how does this sound?

**Comment by EricSimons** [comment]
@alexisvincent I think you're right that bundling is the correct solution here. Even with cache hits, loading the non-dist versions of react/redux from unpkg + booting up takes approx 3-5s after page load for me, which obviously is a bit too long for a production use case. Bundling would drop that number substantially though.

Btw I think that this is an excellent idea regardless of whether it's perfectly suitable for production use cases. In fact, I think the use case is even better for non-production use cases, as this solution would lower the barrier to entry for building side projects, examples, etc - things that people might not normally start building due to the unavoidable upfront cost of rolling their tooling (webpack, etc).

**Comment by alexisvincent** [comment]
@EricSimons Exactly, thats the main target market. We have a web dev course at Uni and last year I saw too many folk get really bad impressions of the web, because of the implicit complexities of the web. I want to have something that I can hand to people and say, here, use this. But that is actually simple, not just artificially so.

I think I way have a rough solution for the `HTTP 1.x` case. There are various more optimized variations of this but I'm going to write down the simplest so that everyone can get the basic idea.

Lets assume the project has just 3 dependencies, `react@15.4.2`, `redux@3.6.0` and `lodash@4.17.2` and imagine for a moment that these projects didn't have transitive dependencies (makes the explanation easier).

We can construct a url, where deps are ordered alphabetically, that tells `unpkg` upfront all the deps you will need, which would look something like this: `https://unpkg.com?bundle=lodash@14.17.2:react@15.4.3:redux@3.6.0`. Upon first load, that file could be requested, `unpkg` could then determine if the incoming request is `HTTP 2 or 1`. 

In the case that it's `HTTP 1`, it could stream the result of the concatenation of `https://unpkg.com/lodash@14.17.2?bundle=system`, `https://unpkg.com/react@15.4.3?bundle=system`, `https://unpkg.com/redux@3.6.0?bundle=system`. We can set the cache timeout to 1 year. So in one request the app will have all the deps it needs.

In the case that it's `HTTP 2`, we can return a null response, with either a header, a cookie or some code, that informs SystemJS that it's on a host that supports http2 (I haven't found a way to do this without making a request), it can then load `https://unpkg.com/lodash@14.17.2?bundle=system` and the others itself. Once again the response could be cached and subsequent loads wouldn't have to ask `unpkg` if the client supports `HTTP 2`. Then if in the meantime they update their browser, the next time the site updates, they can leverage the `HTTP 2` load and caching. We could leverage `server push` in the first request, but I haven't yet figured out a way (without using Service Workers) to inform the server which files the client has so as to avoid pushing deps it has in its `http` cache. This is almost something I wish was fixed by the protocol.

If we get buy-in from the person serving the index file (maybe by providing an express middleware or something), we could optimize this a bit by letting that server inform `SystemJS` whether or not the browser supports `HTTP 2`, it could also inline the `SystemJS` source into the index file, so that SystemJS gets control as soon as possible.

Another variation would be to along with the request for `https://unpkg.com?bundle=lodash@14.17.2:react@15.4.3:redux@3.6.0` to request the others simultaneously and let the `unpkg` server kill those requests, but you then will hit the `max connection limit` and so that would cause a temporary block on the other asset loads.

Assuming a 40ms download time (latency + download) on all files to both servers we get the following approximate first hit load times. Ignoring time taken to parse and load the JS.

**Unoptimized HTTP 1**
40ms (initial index load) + 40ms (SystemJS load and bundle request) = 80ms

**Optimized HTTP 1**
40ms (initial index load + SystemJS load) + 40ms (bundle request) = 80ms

**Unoptimized HTTP 2**
40ms (initial index load) + 40ms (SystemJS load and bundle request) + 40ms (all small bundles) = 120ms

**Optimized HTTP 2**
40ms (initial index load + SystemJS load + protocol hinting) + 40ms (bundle request and small bundle requests) = 80ms

While in the worst case **Unoptimized HTTP 2** is one roundtrip slower then the others, it makes up for that by providing per package caching. Also, `HTTP 2`, can leverage internal cache from visiting other websites that use the same scheme.

Unless @EricSimons can once again shatter my dreams, this could be a viable way forwards. Let me know your thoughts.

**Comment by alexisvincent** [comment]
> Even with cache hits, loading the non-dist versions of react/redux from unpkg + booting up takes approx 3-5s after page load for me, which obviously is a bit too long for a production use case.

@EricSimons Can you explain your experiment again. I didn't quite understand what you tested.

**Comment by alexisvincent** [comment]
@EricSimons Would love to chat regarding your Thinkster project!

**Comment by mjackson** [comment]
Also worthwhile noting in this discussion is that the `<script type="module">` spec supports loading directly from URLs. So e.g. you can do stuff like [this](https://github.com/matthewp/bram-es/blob/master/index.html#L8). /cc @matthewp

**Comment by matthewp** [comment]
A crazy idea that's been in my brain for the last few days that i just wanted to throw out. What if there was a babel plugin that transformed:

```js
import React from 'react';
```

into:

```js
import React from '//unpkg.com/react@1.0.0/es-main.js';
```

assuming that this module had an ES main.  And it did this recursively for all dependencies and their dependencies.

That part is easy, but what if unpkg had a mode that did this for you, ran the code through babel using this plugin (and only this plugin), caching the results of course.  Then essentially all of your code in development mode would be coming from unpkg.

**Comment by mjackson** [comment]
@matthewp It's crazy you said that. I've had literally the exact same thought running through my head this past week.

I was thinking about possibly using a `?module` query string parameter to tell unpkg to run this transform on all your code. Then use `<script type="module">` and away you go!

**Comment by mjackson** [comment]
> If intruducing new API, better make it like polyfill or at leadt compatible.

Yep. A `?module` query parameter would be completely backwards compatible. If you don't have it, you'll get the same file you've always gotten. If you do, party!

**Comment by alexisvincent** [comment]
What ever does get released, its worth coordinating them, so that it doesn't end up confusing folks when a ton of ways of doing something come out of unpkg.

**Comment by mjackson** [comment]
For sure. @matthewp is actually talking about moving in that direction and supporting `<script type="module">` better on the unpkg side, which I believe is the natural way for unpkg to evolve. Sure, we could include better support for bundlers, which I'm also interested in doing. But ultimately I believe the future of unpkg should be as a CDN for individual files.

2 things need to happen in order to make this work well:

- *People need to start publishing packages with ES modules.* Not a lot of people are publishing ES modules to npm right now. It's mostly still CommonJS. But the bundlers themselves are getting better as supporting ES modules out of the box (see Rollup and webpack 2) so I believe that more package authors will start including them. I know we're already supporting them in `react-router`.
- Browsers need to support `<script type="module">`, which Safari TP already does. I'd expect more browsers to include support as the spec formalizes.

Anyway, to avoid hijacking this issue too much I've started a new thread in https://github.com/unpkg/unpkg.com/issues/24

**Comment by abhishiv** [comment]
@alexisvincent I have been working on exactly on the same thing, but have used jspm itself to install packages, upload them to s3, and then extract config. As you can imagine the workflow isn't quite optimized. I'm trying to comeup with a way to use unpkg, and your systemjs-config-builder project.

Since I'm guessing you are quite busy with other projects, would you be kind enough to document which areas need work on. I have a few free days at my disposal, and would love to see unpkg work with systemjs.

**Comment by alexisvincent** [comment]
@abhishiv Interestingly enough I started working on this again last night and this morning and have been working on a large rewrite to support this issue. I don't have a lot of time to put into it so if you're keen to help that would be awesome. 

Theres a few things that need to happen here.

**systemjs-config-builder**
- [ ] support the full npm resolution logic (working on)
- [ ] be able to just process one package
- [ ] support overrides via JSPM overrides registry

**unpkg**
- [ ] generate configs on a package level
- [ ] generate bundles for the packages

**[unpkg-loader](https://github.com/alexisvincent/systemjs-unpkg-loader)**
- [ ] resolve the modules from unpkg via generated configs

The biggest two blockers atm are the unpkg support and systems-config-builder. Of the two, systemjs-config-builder would be the most helpful to receive help with. Specifically the overrides. 

Let me know what you're keen to help with and we can nail this thing.

**Comment by nkbt** [comment]
Please, check out this cool thing made by people from Atlassian and presented last week on our local SydJS, it packages scripts with webpack on the fly into a one working bundle.

https://pkgzip.com

**Comment by alexisvincent** [comment]
@nkbt Very cool! We're wanting to do a very similar thing! Just more customizable and without including all transitive deps. Thanks for the link

**Comment by mjackson** [comment]
@nkbt There's also https://wzrd.in/ which does the same thing, just using Browserify instead of webpack. I think Browserify is a better fit for an auto-bundler service because it works better "out of the box" with most npm modules, whereas webpack usually requires some sort of custom config.

**Comment by nkbt** [comment]
Yeah, I reckon `wzrd` was mentioned in a talk. `unpkg` was also endorsed several rimes =). I hope they will publish the video/slides eventually.

**Comment by alexisvincent** [comment]
I've been working on the config builder and I have a nice new release coming that would support this loader. Once I have that we can look at getting config generation in to unpkg.

**Comment by abhishiv** [comment]
Hey guys

Just wanted to point out the work @cerebral did with Webpack DLL 

[@cerebral/webpack-packager](https://github.com/cerebral/webpack-packager)
[@cerebral/webpack-dll](https://github.com/cerebral/webpack-dll)

[Blog post](https://medium.com/@christianalfoni/webpack-bundling-as-a-service-f902ab1a9f4c
)

It's not SystemJS based, but similar in concept. Right now it seems JSDeliver implements it 
https://cdn.jsdelivr.net/webpack/v1/react-dom@15.5.4+react@15.5.4/dll.js

Would like to know what you guys think of a webpack DLL based approach instead of SystemJS.

**Comment by mjackson** [comment]
@abhishiv Please open a separate issue if you'd like to discuss something other than SystemJS.

**Comment by mjackson** [comment]
I'm closing this repo, so I've migrated the issue to https://github.com/unpkg/unpkg.com/issues/35. Let's continue discussing there. Thanks!
